File: src/BehaviorsSDKNative/Microsoft.Xaml.Interactivity/Interaction.cpp
#include "pch.h"
#include "Interaction.h"
#include "BehaviorCollection.h"

namespace winrt
{
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI::Xaml;
using namespace Microsoft::Xaml::Interactivity;
} // namespace winrt

namespace winrt::Microsoft::Xaml::Interactivity::implementation
{
namespace DependencyProperties
{
static DependencyProperty Behaviors = DependencyProperty::RegisterAttached(
    L"Behaviors",
    xaml_typename<Microsoft::Xaml::Interactivity::BehaviorCollection>(),
    xaml_typename<Microsoft::Xaml::Interactivity::Interaction>(),
    PropertyMetadata(nullptr, {&Interaction::OnBehaviorsChanged}));
}

void Interaction::OnBehaviorsChanged(DependencyObject const &sender, DependencyPropertyChangedEventArgs const &args)
{
  auto oldCollection = args.OldValue().as<Microsoft::Xaml::Interactivity::BehaviorCollection>();
  auto newCollection = args.NewValue().as<Microsoft::Xaml::Interactivity::BehaviorCollection>();

    if (oldCollection == newCollection)
    {
        return;
    }

    if (oldCollection != nullptr && oldCollection.AssociatedObject() != nullptr)
    {
        oldCollection.Detach();
    }

    if (newCollection != nullptr && sender != nullptr)
    {
        newCollection.Attach(sender);
    }
}

DependencyProperty Interaction::BehaviorsProperty()
{
    return DependencyProperties::Behaviors;
}

IIterable<IInspectable> Interaction::ExecuteActions(IInspectable const &sender, ActionCollection const &actions, IInspectable const &parameter)
{
    IVector<IInspectable> results = winrt::single_threaded_vector<IInspectable>();

    if (actions == nullptr || Windows::ApplicationModel::DesignMode::DesignModeEnabled())
    {
        return results;
    }

    for (auto &dependencyObject : actions)
    {
        auto action = dependencyObject.as<IAction>();
        results.Append(action.Execute(sender, parameter));
    }

    return results;
}

Microsoft::Xaml::Interactivity::BehaviorCollection Interaction::GetBehaviors(DependencyObject const &obj)
{
    if (obj == nullptr)
    {
        throw winrt::hresult_invalid_argument(L"obj");
    }

    auto behaviors = obj.GetValue(Interaction::BehaviorsProperty()).as<Microsoft::Xaml::Interactivity::BehaviorCollection>();
    if (behaviors == nullptr)
    {
        behaviors = make<BehaviorCollection>();
        Interaction::SetBehaviors(obj, behaviors);
    }

    return behaviors;
}

void Interaction::SetBehaviors(DependencyObject const &obj, Microsoft::Xaml::Interactivity::BehaviorCollection const &value)
{
    if (obj == nullptr)
    {
        throw winrt::hresult_invalid_argument(L"obj");
    }

    obj.SetValue(Interaction::BehaviorsProperty(), value);
}

} // namespace winrt::Microsoft::Xaml::Interactivity::implementation

File: src/BehaviorsSDKManaged/Microsoft.Xaml.Interactivity.Shared/Interaction.cs
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;

#if WinUI
using Microsoft.UI.Xaml;
#else
using Windows.UI.Xaml;
#endif

namespace Microsoft.Xaml.Interactivity
{
    /// <summary>
    /// Defines a <see cref="BehaviorCollection"/> attached property and provides a method for executing an <seealso cref="ActionCollection"/>.
    /// </summary>
    public sealed class Interaction
    {
        /// <remarks>
        /// CA1053: Static holder types should not have public constructors
        /// </remarks>
        private Interaction()
        {
        }

        /// <summary>
        /// Gets or sets the <see cref="BehaviorCollection"/> associated with a specified object.
        /// </summary>
        [SuppressMessage("Microsoft.Security", "CA2104:DoNotDeclareReadOnlyMutableReferenceTypes")]
        public static readonly DependencyProperty BehaviorsProperty = DependencyProperty.RegisterAttached(
            "Behaviors",
            typeof(BehaviorCollection),
            typeof(Interaction),
            new PropertyMetadata(null, new PropertyChangedCallback(Interaction.OnBehaviorsChanged)));

        /// <summary>
        /// Gets the <see cref="BehaviorCollection"/> associated with a specified object.
        /// </summary>
        /// <param name="obj">The <see cref="Windows.UI.Xaml.DependencyObject"/> from which to retrieve the <see cref="BehaviorCollection"/>.</param>
        /// <returns>A <see cref="BehaviorCollection"/> containing the behaviors associated with the specified object.</returns>
        public static BehaviorCollection GetBehaviors(DependencyObject obj)
        {
            if (obj == null)
            {
                throw new ArgumentNullException(nameof(obj));
            }
            BehaviorCollection behaviorCollection = (BehaviorCollection)obj.GetValue(Interaction.BehaviorsProperty);
            if (behaviorCollection == null)
            {
                behaviorCollection = new BehaviorCollection();
                obj.SetValue(Interaction.BehaviorsProperty, behaviorCollection);

                var frameworkElement = obj as FrameworkElement;

                if (frameworkElement != null)
                {
                    frameworkElement.Loaded -= FrameworkElement_Loaded;
                    frameworkElement.Loaded += FrameworkElement_Loaded;
                    frameworkElement.Unloaded -= FrameworkElement_Unloaded;
                    frameworkElement.Unloaded += FrameworkElement_Unloaded;
                }
            }

            return behaviorCollection;
        }

        /// <summary>
        /// Sets the <see cref="BehaviorCollection"/> associated with a specified object.
        /// </summary>
        /// <param name="obj">The <see cref="Windows.UI.Xaml.DependencyObject"/> on which to set the <see cref="BehaviorCollection"/>.</param>
        /// <param name="value">The <see cref="BehaviorCollection"/> associated with the object.</param>
        public static void SetBehaviors(DependencyObject obj, BehaviorCollection value)
        {
            if (obj == null)
            {
                throw new ArgumentNullException(nameof(obj));
            }
            obj.SetValue(Interaction.BehaviorsProperty, value);
        }

        /// <summary>
        /// Executes all actions in the <see cref="ActionCollection"/> and returns their results.
        /// </summary>
        /// <param name="sender">The <see cref="System.Object"/> which will be passed on to the action.</param>
        /// <param name="actions">The set of actions to execute.</param>
        /// <param name="parameter">The value of this parameter is determined by the calling behavior.</param>
        /// <returns>Returns the results of the actions.</returns>
        public static IEnumerable<object> ExecuteActions(object sender, ActionCollection actions, object parameter)
        {
            List<object> results = new List<object>();

            if (actions == null || Windows.ApplicationModel.DesignMode.DesignModeEnabled)
            {
                return results;
            }

            foreach (DependencyObject dependencyObject in actions)
            {
                IAction action = (IAction)dependencyObject;
                results.Add(action.Execute(sender, parameter));
            }

            return results;
        }

        private static void OnBehaviorsChanged(DependencyObject sender, DependencyPropertyChangedEventArgs args)
        {
            BehaviorCollection oldCollection = (BehaviorCollection)args.OldValue;
            BehaviorCollection newCollection = (BehaviorCollection)args.NewValue;

            if (oldCollection == newCollection)
            {
                return;
            }

            if (oldCollection != null && oldCollection.AssociatedObject != null)
            {
                oldCollection.Detach();
            }

            if (newCollection != null && sender != null)
            {
                newCollection.Attach(sender);
            }
        }

        private static void FrameworkElement_Loaded(object sender, RoutedEventArgs e)
        {
            var d = sender as DependencyObject;

            if (d != null)
            {
                GetBehaviors(d).Attach(d);
            }
        }

        private static void FrameworkElement_Unloaded(object sender, RoutedEventArgs e)
        {
            var d = sender as DependencyObject;

            if (d != null)
            {
                GetBehaviors(d).Detach();
            }
        }
    }
}